src/Modules/OrchardCore.Commerce/Events/InventoryCheckoutEvents.cs

```diff
@@ -0,0 +1,26 @@
+using OrchardCore.Commerce.Abstractions;
+using OrchardCore.Commerce.Abstractions.Abstractions;
+using OrchardCore.Commerce.Abstractions.ViewModels;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Commerce.Events;
+
+public class InventoryCheckoutEvents : ICheckoutEvents
+{
+    private readonly IProductInventoryService _productInventoryService;
+
+    public InventoryCheckoutEvents(
+        IProductInventoryService productInventoryService) => _productInventoryService = productInventoryService;
+
+    public async Task ViewModelCreatedAsync(IList<ShoppingCartLineViewModel> lines, ICheckoutViewModel checkoutViewModel)
+    {
+        var cannotCheckout = await _productInventoryService.VerifyLinesAsync(lines);
+        if (cannotCheckout)
```

Comment:
These could be merged too, the "cannot checkout" information is communicated by the `IsInvalid = true`.
```suggestion
        if (await _productInventoryService.VerifyLinesAsync(lines))
```